### ISSUE TITLE
Weird/Inconsitent Typescript behavior with Svelte libraries

### ISSUE THREAD

**Issue by Shyrro** [withastro/astro#8460]
### Astro Info

```block
Astro                    v3.0.10
Node                     v18.17.1
System                   macOS (arm64)
Package Manager          pnpm
Output                   static
Adapter                  none
Integrations             @astrojs/svelte
```

### Describe the Bug

When using a Svelte library ( Currently the "official" way to build a svelte library is to use `svelte-package` ), types are weirdly inferred, sometimes throwing error on lint but not on build, other times throwing error on build not on lint and sometimes not throwing errors at all and the component props do not have the right typings. 

I thought the issue was coming from a custom library, so i created a reproduction where we have uses with a simple library and other uses with Svelte HeadlessUI.

Here are the weird cases that were observed.

**Case 1 : Every component prop inferred to "string"** 

To test this case, i created a Svelte component library in the same workspace as the Astro app. The library export a simple Svelte Component : 

```html
<script lang="ts">
  type Variant = "test1" | "test2" | "test3";
  export let variant: Variant = "test1";
  export let open: boolean = false;
</script>

<div>
  {variant}
  isOpen : {open}
</div>
```

When using it with Astro, every prop is inferred to string. The types in the bundle seem correct.

Example : 

![Capture d’écran 2023-09-08 à 10 52 05](https://github.com/withastro/astro/assets/10008963/a341f19f-bf3e-4c54-8618-c67b066e114a)

I thought this was the extension at first, but the build doesn't seem to bother either. 

**Case 2: No linting error but build breaks**

To test this, i used the library Svelte Headless UI : https://github.com/rgossiaux/svelte-headlessui 

In some cases, some properties are inferred to `any` by the linter( or the extension ? ). 

In this example, i use the component `Dialog` from Headless UI which has a property `open` that is a `boolean`. In the `.astro` file, this prop is inferred to `any` and i can pass whatever i want. 

![Capture d’écran 2023-09-08 à 10 58 43](https://github.com/withastro/astro/assets/10008963/30271315-ff5b-4ebc-9314-c9de35f7ab8a)

As you can see, i can pass the prop `open` as a string and my VSCode linter doesn't bother because it's a string for him. However, when running the build command for the astro app, it breaks correctly, telling me that `open` is a boolean and cannot accept a string. 

**Case 3: Linter throwing errors, but not the build**

To test this, i used the component `TabGroup` from Svelte Headless UI. This component has a prop `defaultIndex`, which is a number.

In my astro file, i call this component passing a string to the `defaultIndex` prop. This throws an error in the linter. 
Which felt great until the build passed... 

We can also notice that in the image, the case 1 is present with the prop `orientation`, which type is `"vertical" | "horizontal"`, but here is being inferred to `string`.

You can find all the reproductions in the github repo linked in the minimal reproduction. 
The reproduction contains a pnpm workspace with : 

- 1 Svelte lib
- 1 Astro app 

There are scripts in the repo to build each and to build all. 

- build all : `pnpm build`
- build svelte lib : `pnpm build:lib`
- build astro app : `pnpm build:app`

In the `index.astro` there is all the use cases also containing comments as to what doesn't work. 
Hope this provides enough, i am available for give more if needed.

### What's the expected result?

I expect the types to be inferred correctly both within my editor and in the build. 

This also feels inconsistent, i am confused as to what doesn't work here and even if i wanted to contribute to fix this, i wouldn't know where to start. 

I used Headless UI because it's an open source library, but these error also happened with private libraries that i am not able to share, so i don't think it's coming from HeadlessUI, since in a Vite or SvelteKit App these work correctly.

I know Svelte is also peculiar compared to other frameworks in the way it works/compiles so maybe it needs some extra effort for TS...

As said, i am willing to submit a pull request/contribute to this, i would just need some hints as to where i can get started :D 

### Link to Minimal Reproducible Example

https://github.com/Shyrro/astro-svelte-ts-repro

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by Princesseuh** [comment]
It's not that everything is being inferred as string, it cannot find the type so it shows the type of what you pass.

This is ultimately a combination of a fair amount of issues:

- Svelte types are very very complicated and were not intended to be used by other tools (which is fair)
- Our language server does not support Svelte components's .d.ts <https://github.com/withastro/language-tools/issues/425> (maybe it does better now? But I remember those being quite Svelte specific)
- `pnpm` has problems resolving `svelte2tsx` types if it's not hoisted / installed into the user's project

The TL;DR is that, Svelte doesn't make this easy. Importing a direct `.svelte` file will typically work, but setups with barrel imports, `.d.ts` will typically face difficulties. This is especially worse in a monorepo.

> In my astro file, i call this component passing a string to the defaultIndex prop. This throws an error in the linter.
Which felt great until the build passed...

Astro does not type check on build, so it's normal for your build to work even with type errors. See https://docs.astro.build/en/guides/typescript/#type-checking for ways to type check before building.

**Comment by Shyrro** [comment]
Hey @Princesseuh 

Thanks for taking the time to answer. This seems fair, i faced issues with Svelte on other occasions so i totally get what you're saying. 🙏 

The point that still raises a question to me, is that, if astro does not typecheck, why does the build fail in Case 2 ? Case 3 makes more sense

I don't know if there is any explanation to that, this is more to open the discussion and understand what's happening behind the scenes.
